connection_test.go

```diff
@@ -970,3 +970,49 @@ func TestPeerStatusChangeServer(t *testing.T) {
 	})
 	assert.Len(t, changes, 0, "unexpected peer status changes")
 }
+
+func TestContextCanceledOnClientTCPClose(t *testing.T) {
+
+	opts := testutils.NewOpts().DisableLogVerification().NoRelay()
+
+	testutils.WithTestServer(t, opts, func(server *testutils.TestServer) {
+
+		dispatchC := make(chan struct{})
+		callDoneC := make(chan struct{})
+
+		server.RegisterFunc("test", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+			defer close(callDoneC)
+			close(dispatchC)
+			<-ctx.Done()
+			assert.Equal(t, "test-client", CurrentCall(ctx).CallerName(), "expected caller name to be passed through")
+			return &raw.Res{}, nil
+		})
+
+		conn, err := net.Dial("tcp", server.HostPort())
+		assert.NoError(t, err, "tcp connect to server failed")
+
+		err = DoInitHandshake(conn, 1, conn.LocalAddr().String())
```

Comment:
This test is very low level, but doesn't need to be.

Another way to write this test:
 - set up a [frame relay](http://godoc.org/github.com/uber/tchannel-go/testutils#FrameRelay) which will let you close a connection by calling the returned cancel function
 - set up a server that has a handler that just waits for the context `Done` and verifies that the error is `Canceled` once it's complete
 - make a call in a background goroutine with a long timeout (like 10 seconds) 
 - wait for the call on the frame relay, once you see it, close the relay, which will cause all connections to close, and *should* trigger the server context to end
 - verify that the server handler is done (wait for a channcel to close with a timeout that's significantly shorter than the call timeout)

We don't want tests that rely on writing out frames exactly.